src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,753 @@
+    auto permutation = getCanonicalPermutation(groupInfo.group);
+    groupInfo.permutation = otherInfo.permutation = std::move(permutation);
+
+    // Set up `other` to be the representative element of the equivalence class.
+    otherInfo.classInfo.emplace(otherInfo);
+    auto& classInfo = *otherInfo.classInfo;
+
+    // Update both groups to have the initial valid order. Their shapes still
+    // match.
+    classInfo.permute(otherInfo);
+    classInfo.permute(groupInfo);
+
+    // Insert `other` with its new shape. It may end up being joined to another
+    // existing equivalence class, in which case its class info will be cleared
+    // and `group` will subsequently be added to that same existing class, or
+    // alternatively it may be inserted as the representative element of a new
+    // class to which `group` will subsequently be joined.
+    shapesToUpdate.push_back(group);
+    shapesToUpdate.push_back(other);
+  }
+
+  std::vector<size_t>
+  getCanonicalPermutation(const std::vector<HeapType>& types) {
+    // The correctness of this part depends on some interesting properties of
+    // strongly connected graphs with ordered, directed edges. A permutation of
+    // the vertices in a graph is an isomorphism that produces an isomorphic
+    // graph. An automorphism is an isomorphism of a structure onto itself, so
+    // an automorphism of a graph is a permutation of the vertices that does not
+    // change the graph. Permutations can be described in terms of sets of
+    // cycles of elements.
+    //
+    // Theorem 1: All cycles in an automorphism of an SCC are the same size.
+    //
+    //     Proof: By contradiction. Assume there are two cycles of different
+    //     sizes. Because the SCC is fully connected, there is a path from an
+    //     element in the smaller of these cycles to an element in the larger.
+    //     This path must contain an edge between two vertices in cycles of
+    //     different sizes N and M such that N < M. Apply the automorphism N
+    //     times. The source of this edge has been cycled back to its original
+    //     index, but the destination is not yet back at its original index, so
+    //     the edge's destination index is different from its original
+    //     destination index and the permutation we applied was not an
+    //     automorphism after all.
+    //
+    // Corollary 1.1: No nontrivial automorphism of an SCC may have a stationary
+    // element, since either all the cycles have size 1 and the automorphism is
+    // trivial or all the cycles have some other size and there are no
+    // stationary elements.
+    //
+    // Corollary 1.2: No two isomorphic SCCs with the same first element are the
+    // same since no nontrivial automorphism can keep the first element
+    // stationary.
+    //
+    // Theorem 2: SCCs with initial elements that are not in an automorphism
+    // cycle with each other are not automorphic to each other.
```

Comment:
Say I have two SCCs, `(A,B,C,D)` and `(X,Y,Z,W)`. How can I talk about an automorphism cycle between them? Don't I need to have an automorphism for that? Are you talking about all possible automorphisms between them perhaps?

I am also a bit confused by the term "automorphic". "Isomorphic" makes sense, in that it says that two things *can* be mapped to each other,

> Two mathematical structures are isomorphic if an isomorphism exists between them. (wikipedia)

But everything is "automorphic"  in a boring way (everything can be mapped to itself, using the identity, trivially). Indeed, the term "automorphic" doesn't appear on [Wikipedia's page for "Automorphism"](https://en.wikipedia.org/wiki/Automorphism), and searching elsewhere I can't find general use of the term in math (only specific ones). Maybe I'm missing it?

Reply:
> Say I have two SCCs, (A,B,C,D) and (X,Y,Z,W). How can I talk about an automorphism cycle between them?

Yeah, this is pretty much a category error. That's why I think deleting this makes the most sense. PTAL at the commit "delete theorem 2" for the fix.